bot-components/Utils.ml

```diff
@@ -1,139 +1,111 @@
-                    |> List.filter ~f:(fun entry -> not entry.is_directory)
-                    |> List.map ~f:(fun entry ->
-                           (entry, Zip.read_entry zf entry) )
-                  in
-                  Zip.close_in zf ; entries
-                in
-                Ok zip_entries
-              with Zip.Error (zip_name, entry_name, message) ->
-                Error (f "Zip.Error(%s, %s, %s)" zip_name entry_name message) ) )
-  >|= action
-
-(* GitHub specific *)
-
-let project_api_preview_header =
-  [("Accept", "application/vnd.github.inertia-preview+json")]
-
-let app_api_preview_header =
-  [("Accept", "application/vnd.github.machine-man-preview+json")]
-
-let api_json_header = [("Accept", "application/vnd.github+json")]
-
-let github_header bot_info =
-  [("Authorization", "bearer " ^ github_token bot_info)]
-
-let headers_of_list = headers
-
-(* when following a redirect from GitHub to Azure, passing along the
-   Authorization header results in 403 Forbidden.  So we strip the
-   headers when we recurse by default. *)
-let rec client_get ?(follow_redirects = true)
-    ?(include_headers_in_redirects = false) ~user_agent ~headers uri =
-  Client.get ~headers uri
-  >>= fun (resp, body) ->
-  match Response.status resp with
-  | `OK ->
-      Lwt.return_ok body
-  | `Moved_permanently
-  | `Found
-  | `See_other
-  | `Temporary_redirect
-  | `Permanent_redirect
-    when follow_redirects -> (
-      let headers =
-        if include_headers_in_redirects then headers
-        else headers_of_list [] user_agent
+let toml_of_string s = Toml.Parser.(from_string s |> unsafe)
+
+let toml_of_file file_path = Toml.Parser.(from_filename file_path |> unsafe)
+
+let subkey_value toml_table k k' =
+  Toml.Lenses.(get toml_table (key k |-- table |-- key k' |-- string))
+
+let find k toml_table =
+  Toml.Types.Table.find (Toml.Types.Table.Key.of_string k) toml_table
+
+let list_table_keys toml_table =
+  Toml.Types.Table.fold
+    (fun k _ ks -> Toml.Types.Table.Key.to_string k :: ks)
+    toml_table []
+
+let days_elapsed ts =
+  (* Yes, I know this is wrong because of DST and black holes but it should
+     still be correct enough *)
+  Float.to_int ((Unix.time () -. ts) /. (3600. *. 24.))
+
+let rec apply_throttle len action args =
+  if List.is_empty args || len <= 0 then Lwt.return_unit
+  else
+    let args, rem = List.split_n args len in
+    Lwt_list.map_p action args
+    >>= fun ans ->
+    let n = List.count ~f:(fun b -> b) ans in
+    apply_throttle (len - n) action rem
+
+let format_options_for_getopts options =
+  " " ^ options ^ " " |> Str.global_replace (Str.regexp "[\n\r\t]") " "
+
+let getopts options ~opt =
+  String_utils.map_string_matches
+    ~regexp:(f " %s\\(\\.\\|[ =:-]\\|: \\)\\([^ ]+\\) " opt)
+    ~f:(fun () -> Str.matched_group 2 options)
+    options
+
+let getopt options ~opt =
+  options |> getopts ~opt |> List.hd |> Option.value ~default:""
+
+let parse_mappings mappings =
```

Comment:
This function could probably be closer to the others that are related to GitHub/GitLab synchronizations (currently mostly in `Git_utils`).